proposals/0000-warning-pragmas-with-categories.rst

```diff
@@ -0,0 +1,252 @@
+  as an error, while other warnings are unaffected.
+
+* Since multiple warning categories can be applied to an identifier, users may
+  distinguish more finely if needed, for example ``-Werror -Wwarn=warnings:head
+  -Wno-warnings:tail`` could be used in a codebase where most warnings are
+  considered errors, except uses of ``head`` are merely warned about, and uses
+  of ``tail`` are not mentioned.
+
+
+Proposed Change Specification
+-----------------------------
+
+The `WARNING and DEPRECATED pragmas
+<https://downloads.haskell.org/ghc/9.4.1/docs/users_guide/exts/pragmas.html#warning-deprecated-pragma>`_
+may be immediately followed by a comma-separated list of warning categories in
+square brackets. Each warning category is an identifier beginning with a letter
+and followed by valid identifier characters or dashes.
+
+More formally, the grammar of declarations is extended as follows:
+
+============  =  =====================================================================
+*decl*        →  ``{-# WARNING`` [*categories*] *things* *strings* ``#-}``
+*decl*        →  ``{-# DEPRECATED`` [*categories*] *things* *strings* ``#-}``
+
+*categories*  →  ``[`` *category1*, ..., *categoryN* ``]``
+*category*    →  (*small* | *large*) { *small* | *large* | *digit* | ``'`` | ``-`` }
+
+*things*      →  *thing1*, ..., *thingN*
+*thing*       →  *varid* | *conid*
+*strings*     →  *string* | ``[`` *string1*, ..., *stringN* ``]``
+============  =  =====================================================================
+
+The list of categories can be omitted entirely, so this subsumes the existing
+syntax for ``WARNING`` and ``DEPRECATED`` pragmas.  That is, [*categories*]
+means that the presence of the *categories* non-terminal is optional; if present
+it is a list in square brackets (hence the use of typewriter font ``[``
+... ``]`` in the definition of the *categories* production).
+
+The *category* non-terminal subsumes both *varid* and *conid*, so it is always
+possible to use the name of the thing to which a warning is being attached as
+one of the categories.  The dash character (``-``) is permitted as a character
+in addition to identifier characters, since dashes are frequently used in
+warning names.
+
+There is no change to the existing rules for when ``WARNING`` or ``DEPRECATED``
+pragmas give rise to warnings, except that individual warning categories may be
+enabled or disabled using new ``-Wwarnings:<category>`` or
+``-Wno-warnings:<category>`` options, and their priority may be controlled using
+the ``-Werror=warnings:<category>`` or ``-Wwarn=warnings:<category>`` options.
+Here ``<category>`` represents the name of a *category* according to the grammar
+above.  The ``-Wwarnings-deprecations`` warning flag (and its synonym
+``-Wdeprecations``) controls the display of all warnings from ``WARNING`` or
+``DEPRECATED`` pragmas, regardless of category.  The command-line flags are
```

Comment:
I'd have expected that no category now defaults to “`deprecation`”, and that flag still affects only that category - preserves the existing semantics, without making a flag with deprecation in the name disabling unrelated warnings (that happen to be specified with a related tool)

Reply:
Hmm, I quite like this idea. In particular I think we could say:
 * `{- # WARNING ... #-}` or `{- # DEPRECATED ... #-}` without a category list is a synonym for `{-# WARNING [deprecation] ... #-}` ( and perhaps `DEPRECATED` cannot have a category list)
 * `-Wwarnings` (new) enables all warnings from `WARNING` pragmas regardless of category
 * `-Wwarnings-deprecations` and `-Wdeprecations` become synonyms for `-Wwarnings:deprecation`

I'd vaguely worried about backwards compatibility here, but I guess the only potential surprise is that if a dependency starts using a new category, a downstream library that sets `-Wno-warnings-deprecations` or `-Wno-deprecations` will start seeing the warning until it sets -`Wno-warnings`. That seems okay.

Reply:
Strictly speaking we could do without it. But we have `-Wwarnings-deprecations` at the moment, and given all the flexibility to configure warnings it would seem odd if there wasn't a way to say "don't show me any messages from `WARNING` pragmas" without disabling all built-in warnings as well.

Reply:
I'd say users don't care too much if a warning comes from a pragma or some other source (built in? Plugin?), and would design the feature from that point of view. 

But it's not a big deal to add this warning category group, either, if you think it's useful.